packages/remark-stringify/types/test.ts

```diff
@@ -0,0 +1,50 @@
+import remarkStringify = require('remark-stringify')
+import unified = require('unified')
+import {Node, Parent} from 'unist'
+
+const inferredStringifyOptions = {
+  gfm: true,
+  fences: true,
+  incrementListMarker: false
+}
+
+unified().use(remarkStringify, inferredStringifyOptions)
+
+// These cannot be automatically inferred by TypeScript
+const nonInferredStringifyOptions: Partial<
+  remarkStringify.RemarkStringifyOptions
+> = {
+  fence: '~',
+  bullet: '+',
+  listItemIndent: 'tab',
+  rule: '-',
+  strong: '_',
+  emphasis: '*'
+}
+
+unified().use(remarkStringify, nonInferredStringifyOptions)
+
+const badStringifyOptions = {
+  gfm: 'true'
+}
+
+// $ExpectError
+unified().use(remarkStringify, badStringifyOptions)
+
+function gap(this: unified.Processor) {
+  const Compiler = this.Compiler as typeof remarkStringify.Compiler
+  const visitors = Compiler.prototype.visitors
+  const original = visitors.heading
+
+  visitors.heading = heading
+
+  function heading(
+    this: unified.Processor,
+    node: Node & {depth: number},
+    parent?: Parent
+  ) {
+    return (node.depth === 2 ? '\n' : '') + original.apply(this, [node, parent])
```

Comment:
line 43 and 46 seem to indicate that plugins may have trouble accessing properties added by remark.
Using plain `Node` it gets the error.

```
ERROR: 46:13  strict-comparisons  Cannot compare type 'object' to type 'number
```

When extending `Node` it gets the error

```
Type '(this: Processor, node: Node & { depth: number; }, parent?: Parent | undefined) => string' is not assignable to type 'Visitor'.
  Types of parameters 'node' and 'node' are incompatible.
    Type 'Node' is not assignable to type 'Node & { depth: number; }'.
      Property 'depth' is missing in type 'Node' but required in type '{ depth: number; }'.
```

Thoughts on how different `Node` types from `remark`, `rehype`, `retext`, and `redot` could be supporter? :thought_balloon:

Reply:
One option would be a generic.
Something like

```ts
  interface Transformer<I extends Node = Node, O extends Node = Node> {
    (
      node: I,
      file: VFileCompatible,
      next?: (error: Error | null, tree: O, file: VFile) => {}
    ): Error | O | Promise<O>
  }
```

Which would guarantee that the input and the output are a super type of a unist `Node`.
But could not guarantee that type `I` matches what the previous plugin output, meaning it would be possible for JavaScript errors to slip through the typings.

---

Another would be to add a generic to Processor, and set that generic to a union of all types for the give language.
For example
```ts
type AllUnifiedTypes = Node | Parent
const unifiedProcessor = unified<Settings, AllUnifiedTypes>

type AllRemarkTypes = List | ListItem | Table | ...
const remarkProcessor = unified<Settings, AllRemarkTypes>
```

But then there would be the question of how do plugins participate in the supported node typings?
And how are transitions between languages, like mdast to hast, handled?

---

Other ideas?

Reply:
> Which would guarantee that the input and the output are a super type of a unist Node.
But could not guarantee that type I matches what the previous plugin output, meaning it would be possible for JavaScript errors to slip through the typings.

Looks awesome to me. 💯

Reply:
I don't think the another idea would work with plugin adding `AllUnifiedTypes`... The type can not be extended from plugins, unified or remark. So adopters should define their own `AllUnifiedTypes` based on plugins which they are using.

Reply:
@ChristianMurphy @Rokt33r What’s left to do for this PR to land?

Reply:
The main item is getting this test to pass.

The issue is currently the typing expects `visitor` to be passed a exactly a unist `Node`, when it may be a super type of `Node`, in this case a `HeadingNode` which also includes the `depth` property.

There is still some ongoing discussion into the approach.
The outcome of the discussion will likely also need to be applied to the `Transform` type in unified https://github.com/unifiedjs/unified/blob/e408747ec44fb541081e020a3842212e6a60f485/types/index.d.ts#L310-L316
Which similarly may be a super type of `Node`.

Reply:
If it sticks for too long, we have the option to put off adding visitor type.
Remove the test and that part of the typing, and add it in a follow up PR.

Reply:
Thabks for clarifying! The thread is becoming long so it's hard to know what's being discussed! I'd prefer to get *a* version out, so we can test this is the wild, and push the other breaking changes as well, and then make it perfect later. Could that work (without more breaking typescript changes soon?)

Reply:
Sure,
9011bf8891cf47c210509c4443baf1f38ff5b26f resolves the test issue, and adds a `FIXME` comment for what should be updated.
This is ready for a final review.

Reply:
Opened https://github.com/remarkjs/remark/issues/426 to continue the discussion on typing `visitors` marking this as resolved.